libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpNetworkTransport.kt

```diff
@@ -52,47 +55,86 @@ private constructor(
   ): Flow<ApolloResponse<D>> {
     return flow {
       val millisStart = currentTimeMillis()
-      val httpResponse = DefaultHttpInterceptorChain(
-          interceptors = interceptors + engineInterceptor,
-          index = 0
-      ).proceed(httpRequest)
-
-      if (httpResponse.statusCode !in 200..299) {
-        val maybeBody = if (exposeErrorBody) {
-          httpResponse.body
-        } else {
-          httpResponse.body?.close()
-          null
-        }
-        throw ApolloHttpException(
-            statusCode = httpResponse.statusCode,
-            headers = httpResponse.headers,
-            body = maybeBody,
-            message = "Http request failed with status code `${httpResponse.statusCode}`"
-        )
+      var apolloException: ApolloException? = null
+      val httpResponse: HttpResponse? = try {
+        DefaultHttpInterceptorChain(
+            interceptors = interceptors + engineInterceptor,
+            index = 0
+        ).proceed(httpRequest)
+      } catch (e: ApolloException) {
+        apolloException = e
+        null
       }
 
-      // When using @defer, the response contains multiple parts, using the multipart content type.
-      // See https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md
-      if (httpResponse.isMultipart) {
-        emitAll(
-            multipleResponses(request.operation, customScalarAdapters, httpResponse)
-                .map { it.withHttpInfo(request.requestUuid, httpResponse, millisStart) }
-        )
-      } else {
-        emit(
-            singleResponse(request.operation, customScalarAdapters, httpResponse)
-                .withHttpInfo(request.requestUuid, httpResponse, millisStart)
-        )
+      emitAll(when {
+        httpResponse == null -> {
+          errorResponse(request.operation, apolloException)
+        }
+
+        httpResponse.statusCode !in 200..299 -> {
+          errorResponse(request.operation, httpResponse)
+        }
+
+        // When using @defer, the response contains multiple parts, using the multipart content type.
+        // See https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md
+        httpResponse.isMultipart -> {
+          multipleResponses(request.operation, customScalarAdapters, httpResponse)
+        }
+
+        else -> {
+          singleResponse(request.operation, customScalarAdapters, httpResponse)
+        }
       }
+          .map {
+            it.withHttpInfo(request.requestUuid, httpResponse, millisStart)
+          }
+      )
     }
+        .catch { e ->
+          if (e is ApolloException) {
+            emitAll(errorResponse(request.operation, e))
+          } else {
+            throw e
+          }
```

Comment:
Can we remove this `catch` and instead only try/catch around to throwing methods? `parseJsonResponse` here I believe?

Reply:
Also add a comment to `parseJsonResponse`:

```kotlin
 * @throws IOException if reading [jsonReader] fails
 * @throws JsonEncodingException if the data is not valid json
 * @throws JsonDataException if the data is not of the expected type
 * @throws JsonParseException for other cases  
```